### DISCUSSION TITLE
Split "Fabric Mods" datapack

### DISCUSSION THREAD

**Discussion by diskria** [FabricMC/fabric-api#2627] [Mod Dev Support] [answered]
I have 2 datapacks with achievements, in which the first datapack is the base, and the second overrides some of the achievements from the base (required for hardcore). How do I add both of these datapacks to the data folder so that the "Fabric mods" datapack is split into two datapacks that I can further manage from the code? The datapack for hardcore should automatically turn on and off before world creation, depending on whether the player has selected hardcore mode or not.  (this is roughly how I imagine it to be a mixin)

```
@Mixin(CreateWorldScreen.class)
public class CreateWorldScreenMixin {

    @Shadow
    public boolean hardcore;

    @Shadow
    protected DataPackSettings dataPackSettings;

    @Inject(method = "createLevel", at = @At("HEAD"))
    public void createLevelInject(CallbackInfo ci) {
        // hardcore datapack is disabled by default
        if (hardcore) {
            List<String> enabled = new ArrayList<>(dataPackSettings.getEnabled());
            enabled.add("datapack_for_hardcore");
            dataPackSettings = new DataPackSettings(enabled, new ArrayList<>());
        }
    }
}
```

**Comment by warjort** [comment]
That sounds complicated.
The datapack profiles are loaded when the new world screen is created.
You would need to find a way to reload them when the user toggles the game difficulty.
You can see how fabric modifies this code here:
https://github.com/FabricMC/fabric/blob/1.19.2/fabric-resource-loader-v0/src/client/java/net/fabricmc/fabric/mixin/resource/loader/client/CreateWorldScreenMixin.java

It also wouldn't take into account if the user changes the difficulty in game.

Isn't there another direction you could take?
Maybe doing something with the advancements themselves.
I don't know much about advancements, but I do know you can write and register your own criteria and conditions. 
Maybe this could include a configurable check of the player difficulty?

**Reply by diskria** [reply]
Sorry for the long answer, at that time I stopped studying this question. 

No, in the achievements themselves, I do not want to change anything. the fact is that at least 3 more datapacks are planned, which will also be included when creating the world, depending on the selected settings.

In 1.19.3 Mojang added built-in datapacks and I came back to this question again. There are 2 of these datapacks (bundles and 1.20 content) which probably simplifies my task, since there are also 2 of them and they are somehow sewn into the minecraft code.

I am currently studying this issue and I will be glad to any advice, so if you are still interested in this, tell me.

**Reply by diskria** [reply]
![idea64_0sKViT6Tpi](https://user-images.githubusercontent.com/96978370/209942193-b615fa6b-b59c-4e8e-90de-fb918e9f3456.png)

**Reply by diskria** [reply]
if you imagine that this is a mod, then this is what I need. Two different datapacks that can be selectively enabled, but so that they are in the same mod

**Reply by diskria** [reply]
I really don't understand why there is no such thing in fabric. why combine mod resources into a single datapack. there is no possibility to set the order of datapacks, it is impossible to selectively turn off some...

**Comment by apple502j** [comment, marked as answer]
You probably want `ResourceManagerHelper#registerBuiltinResourcePack`.

**Reply by diskria** [reply]
this is amazing, thanks a lot! exactly what I need :)
